app.py

```diff
@@ -103,33 +106,71 @@
 
 def graph_refresh_solver() -> None:
     """Schedule refresh for packages that are not yet analyzed by solver."""
+    if _THOTH_GRAPH_REFRESH_EAGER_STOP:
+        _LOGGER.info(
+            "Eager stop of scheduling new solver runs for unsolved package versions, packages scheduled: %d",
+            _THOTH_GRAPH_REFRESH_EAGER_STOP,
+        )
+
     indexes = list(_GRAPH_DB.get_python_package_index_urls())
 
     packages = []
     # Iterate over all registered solvers and gather packages which were not solved by them.
     for solver_name in _OPENSHIFT.get_solver_names():
         _LOGGER.info("Checking unsolved packages for solver %r", solver_name)
-        for package_name, version in _GRAPH_DB.retrieve_unsolved_python_packages(solver_name):
-            _LOGGER.info(f"Adding new package %r in version %r", package_name, version)
+        solver_info = _GRAPH_DB.parse_python_solver_name(solver_name)
+        for package_name, version, index_url in _GRAPH_DB.get_unsolved_python_package_versions(
+            os_name=solver_info["os_name"],
+            os_version=solver_info["os_version"],
+            python_version=solver_info["python_version"],
+            count=_THOTH_GRAPH_REFRESH_EAGER_STOP
+        ):
+            _LOGGER.info(
+                f"Adding new package %r in version %r on index %r",
+                package_name,
+                version,
+                index_url,
+            )
             _METRIC_PACKAGES_ADDED.inc()
-            packages.append((package_name, version, solver_name))
+            packages.append((package_name, version, index_url, solver_name))
 
     if not packages:
         _LOGGER.info("No unsolved packages found")
         return
 
-    count = 0
-    for package_name, package_version, solver_name in packages:
-        for index_url in indexes:
-            if _GRAPH_DB.python_package_version_exists(package_name, package_version, index_url, solver_name):
-                _LOGGER.debug(
-                    "Not scheduling solver %r as package %r in version %r was already solved for index %r",
+    for package_name, package_version, index_url, solver_name in packages:
+        if not index_url:
```

Comment:
This is copy-paste. To reduce source code, can we do something like:

indexes = [index_url] or indexes

and then just iterate over indexes (so the latter for-loop which is identical is removed)?

Reply:
Where do we use the indexes in _Openshift.schedule_solver()? https://github.com/thoth-station/common/blob/19ca27e18eb4b65ba15eaf6d6d28ace2e13d4a88/thoth/common/openshift.py#L826